File: Scripts/Python/shift.py
# Shift current selection by given x y amounts using optional mode.
# Author: Andrew Trevorrow (andrew@trevorrow.com), June 2006.
# Updated to use exit command, Nov 2006.
# Updated to check for bounded grid, Oct 2010.

from glife import validint, inside
import golly as g

selrect = g.getselrect()
if len(selrect) == 0: g.exit("There is no selection.")

# use same file name as in shift.lua
INIFileName = g.getdir("data") + "shift.ini"
oldparams = "0 0 or"
try:
    f = open(INIFileName, 'r')
    oldparams = f.readline()
    f.close()
except:
    # should only happen 1st time (INIFileName doesn't exist)
    pass

answer = g.getstring("Enter x y shift amounts and an optional mode\n" +
                     "(valid modes are copy/or/xor, default is or):",
                     oldparams, "Shift selection")
xym = answer.split()

# extract x and y amounts
if len(xym) == 0: g.exit()
if len(xym) == 1: g.exit("Supply x and y amounts separated by a space.")
if not validint(xym[0]): g.exit("Bad x value: " + xym[0])
if not validint(xym[1]): g.exit("Bad y value: " + xym[1])
x = int(xym[0])
y = int(xym[1])

# extract optional mode
if len(xym) > 2:
    mode = xym[2].lower()
    if mode=="c": mode="copy"
    if mode=="o": mode="or"
    if mode=="x": mode="xor"
    if not (mode == "copy" or mode == "or" or mode == "xor"):
        g.exit("Unknown mode: " + xym[2] + " (must be copy/or/xor)")
else:
    mode = "or"

# given parameters are valid so save them for next run
try:
    f = open(INIFileName, 'w')
    f.write(answer)
    f.close()
except:
    g.warn("Unable to save given parameters in file:\n" + INIFileName)

# abort shift if the new selection would be outside a bounded grid
if g.getwidth() > 0:
    gridl = -int(g.getwidth()/2)
    gridr = gridl + g.getwidth() - 1
    newl = selrect[0] + x
    newr = newl + selrect[2] - 1
    if newl < gridl or newr > gridr:
        g.exit("New selection would be outside grid.")
if g.getheight() > 0:
    gridt = -int(g.getheight()/2)
    gridb = gridt + g.getheight() - 1
    newt = selrect[1] + y
    newb = newt + selrect[3] - 1
    if newt < gridt or newb > gridb:
        g.exit("New selection would be outside grid.")

# do the shift by cutting the current selection and pasting it into
# the new position without changing the current clipboard pattern
selcells = g.getcells(selrect)
g.clear(inside)
selrect[0] += x
selrect[1] += y
g.select(selrect)
if mode == "copy":
    g.clear(inside)
g.putcells(selcells, x, y, 1, 0, 0, 1, mode)

if not g.visrect(selrect): g.fitsel()

File: Scripts/Lua/shift.lua
-- Shift current selection by given x y amounts using optional mode.
-- Author: Andrew Trevorrow (andrew@trevorrow.com), Mar 2016.

local g = golly()
local gp = require "gplus"

local selrect = g.getselrect()
if #selrect == 0 then g.exit("There is no selection.") end

-- use same file name as in shift.py
local inifilename = g.getdir("data").."shift.ini"
local oldparams = "0 0 or"
local f = io.open(inifilename, "r")
if f then
    -- get the parameters given last time
    oldparams = f:read("*l") or ""
    f:close()
end

local s = g.getstring("Enter x y shift amounts and an optional mode\n"..
                      "(valid modes are copy/or/xor, default is or):",
                      oldparams, "Shift selection")
local x, y, mode = gp.split(s)

-- check x and y
if y == nil then g.exit("Enter x and y amounts separated by a space.") end
if not gp.validint(x) then g.exit("Bad x value: "..x) end
if not gp.validint(y) then g.exit("Bad y value: "..y) end
x = tonumber(x)
y = tonumber(y)

-- check optional mode
if mode == nil then
    mode = "or"
else
    mode = string.lower(mode)
    if mode == "c" then mode = "copy" end
    if mode == "o" then mode = "or" end
    if mode == "x" then mode = "xor" end
    if not (mode == "copy" or mode == "or" or mode == "xor") then
        g.exit("Unknown mode: "..mode.." (must be copy/or/xor)")
    end
end

-- given parameters are valid so save them for next run
f = io.open(inifilename, "w")
if f then
    f:write(s)
    f:close()
end

-- abort shift if the new selection would be outside a bounded grid
if g.getwidth() > 0 then
    local gridl = -g.getwidth()/2
    local gridr = gridl + g.getwidth() - 1
    local newl = selrect[1] + x
    local newr = newl + selrect[3] - 1
    if newl < gridl or newr > gridr then
        g.exit("New selection would be outside grid.")
    end
end
if g.getheight() > 0 then
    local gridt = -g.getheight()/2
    local gridb = gridt + g.getheight() - 1
    local newt = selrect[2] + y
    local newb = newt + selrect[4] - 1
    if newt < gridt or newb > gridb then
        g.exit("New selection would be outside grid.")
    end
end

-- do the shift by cutting the current selection and pasting it into
-- the new position without changing the current clipboard pattern
local selcells = g.getcells(selrect)
g.clear(0)
selrect[1] = selrect[1] + x
selrect[2] = selrect[2] + y
g.select(selrect)
if mode == "copy" then g.clear(0) end
g.putcells(selcells, x, y, 1, 0, 0, 1, mode)

if not g.visrect(selrect) then g.fitsel() end
